controllers/PlagiarismIthenticateActionHandler.inc.php

```diff
@@ -0,0 +1,385 @@
+		);
+
+		$locale = $ithenticate
+			->setApplicableEulaVersion($submission->getData('ithenticateEulaVersion'))
+			->getApplicableLocale(
+				collect([$submission->getData("locale")])
+					->merge($user->getData("locales"))
+					->merge([$context->getPrimaryLocale(), $site->getPrimaryLocale()])
+					->unique()
+					->toArray()
+			);
+
+		$viewerUrl = $ithenticate->createViewerLaunchUrl(
+			$submissionFile->getData('ithenticateId'),
+			$user,
+			$locale,
+			static::$_plugin->getSubmitterPermission($context, $user),
+			(bool)static::$_plugin->getSimilarityConfigSettings($context, 'allowViewerUpdate')
+		);
+
+		if (!$viewerUrl) {
+			return $request->redirect(
+				null,
+				'user',
+				'authorizationDenied',
+				null,
+				['message' => 'plugins.generic.plagiarism.action.launchViewer.error']
+			);
+		}
+
+		return $request->redirectUrl($viewerUrl);
+	}
+
+	/**
+	 * Schedule the similarity report generate process at iThenticate services's end
+	 *
+	 * @param array $args
+	 * @param Request $request
+	 */
+    public function scheduleSimilarityReport($args, $request) {
+
+		$context = $request->getContext();
+
+		/** @var SubmissionFileDAO $submissionFileDao */
+		$submissionFileDao = DAORegistry::getDAO('SubmissionFileDAO');
+		$submissionFile = $submissionFileDao->getById($args['submissionFileId']);
+
+		/** @var \IThenticate $ithenticate */
+		$ithenticate = static::$_plugin->initIthenticate(
+			...static::$_plugin->getServiceAccess($context)
+		);
+
+		$scheduleSimilarityReport = $ithenticate->scheduleSimilarityReportGenerationProcess(
+			$submissionFile->getData('ithenticateId'),
+			static::$_plugin->getSimilarityConfigSettings($context)
+		);
+
+		if (!$scheduleSimilarityReport) {
+			static::$_plugin->sendErrorMessage("Failed to schedule the similarity report generation process for iThenticate submission id " . $submissionFile->getData('ithenticateId'), $submissionFile->getData('submissionId'));
+			$this->generateUserNotification(
+				$request,
+				NOTIFICATION_TYPE_ERROR, 
+				__('plugins.generic.plagiarism.action.scheduleSimilarityReport.error')
+			);
+			return $this->triggerDataChangedEvent($submissionFile);
+		}
+
+		$submissionFile->setData('ithenticateSimilarityScheduled', 1);
+		$submissionFileDao->updateObject($submissionFile);
+
+		$this->generateUserNotification(
+			$request,
+			NOTIFICATION_TYPE_SUCCESS, 
+			__('plugins.generic.plagiarism.action.scheduleSimilarityReport.success')
+		);
+		
+		return $this->triggerDataChangedEvent($submissionFile);
+    }
+
+	/**
+	 * Refresh the submission's similarity score result
+	 *
+	 * @param array $args
+	 * @param Request $request
+	 */
+    public function refreshSimilarityResult($args, $request) {
+		$context = $request->getContext();
+
+		/** @var SubmissionFileDAO $submissionFileDao */
+		$submissionFileDao = DAORegistry::getDAO('SubmissionFileDAO');
+		$submissionFile = $submissionFileDao->getById($args['submissionFileId']);
```

Comment:
I won't flag all of these, so it's worth a thorough review, but this is also dangerous.